src/configuration/configuration.ts

```diff
@@ -0,0 +1,280 @@
+    return vimOptionInCode !== undefined ? vimOptionInCode : this.getInsertSpaccesFromCode();
+  }
+
+  initialize() {
+    super.initialize();
+    const options = vscode.window.activeTextEditor.options;
+    options.insertSpaces = this.value as boolean;
+    vscode.window.activeTextEditor.options = options;
+  }
+
+  public get value() {
+    if (this.source === OptionValueSource.Runtime) {
+      return this._value;
+    } else {
+      return this.getInsertSpaccesFromCode();
+    }
+  }
+
+  public set value(value: OptionValue) {
+    super.value = value;
+    const oldOptions = vscode.window.activeTextEditor.options;
+    oldOptions.insertSpaces = <boolean> value;
+    vscode.window.activeTextEditor.options = oldOptions;
+  }
+
+  private getInsertSpaccesFromCode(): boolean {
+    if (vscode.window.activeTextEditor) {
+      return vscode.window.activeTextEditor.options.insertSpaces! as boolean;
+    } else {
+      return vscode.workspace.getConfiguration("editor").get("insertSpaces", this.defaultValue);
+    }
+  }
+}
+
+@RegisterOption
+class OptionIskeyword extends BaseOptionDetails {
+  id = "iskeyword";
+  defaultValue = "/\\()\"':,.;<>~!@#$%^&*|+=[]{}`?-";
+
+  fetchCodeConfiguration() {
+    const vimOptionInCode = vscode.workspace.getConfiguration("vim").get<boolean>(this.id, undefined);
+    return vimOptionInCode !== undefined ?
+      vimOptionInCode :
+      vscode.workspace.getConfiguration("editor").get("wordSeparators", this.defaultValue);
+  }
+
+  initialize() {
+    super.initialize();
+    // TODO: modify Code's wordSeparators configuration.
+  }
+}
+
+/**
+ * Vim option override sequence.
+ * 1. `:set {option}` on the fly
+ * 2. VS Code Vim configuration section && overlap configurations between Code and Vim
+ * 3. TODO .vimrc.
+ * 4. VSCodeVim flavored Vim option default values
+ */
+export class Configuration {
+  private static _instance: Configuration | null;
+
+  constructor() {
+    // TODO: Read the real Vim config and override existing configration.
+    // Besides, vimrc settings' priority should be between Default and User Settings.
+  }
+
+  public static getInstance(): Configuration {
+    if (Configuration._instance == null) {
+      Configuration._instance = new Configuration();
+    }
+
+    return Configuration._instance;
+  }
+
+  set(option: string, value: OptionValue, source?: OptionValueSource): void {
+    let targetOption = OptionMap.allOptions[option];
+
+    if (!targetOption) {
+      return;
+    }
+
+    if (value === targetOption.value) {
+      targetOption.source = OptionValueSource.Runtime;
+      return;
+    }
+
+    targetOption.value = value;
+  }
+
+  /**
+   * Return a value of option from vim settings.
+   * @param vim option.
+   * @param defaultValue A value should be returned when no value could be found, is `undefined`.
+   */
+  get(option: string, defaultValue?: OptionValue): OptionValue | undefined {
```

Comment:
I don't actually like this API. I would prefer to have every setting listed off sort of like so

```
class Configuration {
    tabstop: number;
    smartcase: boolean;
    ignorecase: boolean;
}
```

etc.

The advantages here are 

1. All settings accesses will be strongly typed
2. Autocomplete will work for our settings